src/functional.js

```diff
@@ -177,3 +180,53 @@ export const asyncApplySpec =
   (spec) =>
   (...args) =>
     asyncMapObjectTerminals(applyTo(...args))(spec);
+
+export const product = reduce(
+  (a, b) => a.flatMap((x) => b.map((y) => [...x, y])),
+  [[]]
+);
+
+export const sideEffect = (f) => (x) => {
```

Comment:
https://ramdajs.com/docs/#tap ?

Reply:
I'll give you the story behind rewriting some of these:
It started because ramda's naming sucks.
Then the choice to avoid aliasing was made so that we have more control and let the functions do less stuff.
Ramda functions do quite a bit (take a look at https://github.com/ramda/ramda/blob/v0.27.0/source/tap.js), namely currying and transducer support (with a specific implementation).
Nowadays I try to avoid currying, I prefer knowing exactly what the signature and having a single usage. In python this also interfered with both optimizaiton and async support, which we hope to add to gamlajs when the need arises.